Modify S3457: Migrate to LayC - `printf`-style format string are misused (#3305)
SonarSource/rspec: rules/S3457/csharp/rule.adoc

== Why is this an issue?

A [composite format string](https://learn.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting) is a string that contains placeholders, represented by indices inside curly braces "{0}", "{1}", etc. These placeholders are replaced by values when the string is printed or logged.

Because composite format strings are interpreted at runtime, rather than validated by the compiler, they can contain errors that lead to unexpected behaviors or runtime errors.

> This rule validates the correspondence between arguments and composite formats when calling the methods of `String.Format`, `StringBuilder.AppendFormat`, `Console.Write`, `Console.WriteLine`, `TextWriter.Write`, `TextWriter.WriteLine`, `Debug.WriteLine(String, Object[])`, `Trace.TraceError(String, Object[])`, `Trace.TraceInformation(String, Object[])`, `Trace.TraceWarning(String, Object[])` and `TraceSource.TraceInformation(String, Object[])`.

Reviewer: Reorganize this into bullet points with links to the documentation of each method on _learn.microsoft.com_, the same way you did with the Java spec.

== Why is this an issue?

> A composite format string is a string that contains placeholders, represented by indices inside curly braces "{0}", "{1}", etc. These placeholders are replaced by values when the string is printed or logged.

Reviewer: ```suggestion
A [composite format string](https://learn.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting) is a string that contains placeholders, represented by indices inside curly braces "{0}", "{1}", etc. These placeholders are replaced by values when the string is printed or logged.
```

Because composite format strings are interpreted at runtime, rather than validated by the compiler, they can contain errors that lead to unexpected behaviors or runtime errors.

This rule statically validates the correspondence between arguments and composite formats when calling the methods of `String.Format`, `StringBuilder.AppendFormat`, `Console.Write`, `Console.WriteLine`, `TextWriter.Write`, `TextWriter.WriteLine`, `Debug.WriteLine(String, Object[])`, `Trace.TraceError(String, Object[])`, `Trace.TraceInformation(String, Object[])`, `Trace.TraceWarning(String, Object[])` and `TraceSource.TraceInformation(String, Object[])`.

=== Exceptions

> * No issue is raised if the format string is not a `const`.

Reviewer: ```suggestion
* No issue is raised if the format string is not a string literal, but comes from a variable.
```

[source,csharp]
----
var pattern = "{0} {1} {2}";

> var res = string.Format(pattern, 1, 2);

Reviewer: ```suggestion
var res = string.Format(pattern, 1, 2); // Incorrect, but the analyzer doesn't raise any warnings here
```

----

* No issue is raised if the argument is not an inline created array.

[source,csharp]
----
var array = new int[] {};
var res = string.Format("{0} {1}", array); // Compliant we don't know the size of the array
----

* This rule doesn't check whether the format specifier (defined after the `:`) is actually valid.

== How to fix it

A composite format string contains placeholders, replaced by values when the string is printed or logged. Mismatch in the format specifiers and the arguments provided can lead to incorrect strings being created.

To avoid issues, a developer should ensure that the provided arguments match format specifiers.

> 

Reviewer: Add the following:

```
Use https://learn.microsoft.com/en-us/dotnet/csharp/tutorials/string-interpolation[string interpolation] when possible. 

Instead of 

[source,csharp]
----
string str = string.Format("Hello {0} {1}!", firstName, lastName); 
----

use

[source,csharp]
----
string str = $"Hello {firstName} {lastName}!"; 
----

With string interpolation:

* the arguments are evulated/validated at compilation time rather than runtime
* modern code editors provide auto-completion when typing the interpolation expression
```
